### PR TITLE
Improve efficiency of private send code

### PR DESCRIPTION
**Pull request by tgflynn**

### CODE REVIEW

File: src/masternodeman.h [hunk 1]
```diff
@@ -106,7 +106,7 @@ class CMasternodeMan
     CMasternode* GetNextMasternodeInQueueForPayment(int nBlockHeight, bool fFilterSigTime, int& nCount);
 
     /// Find a random entry
-    CMasternode* FindRandomNotInVec(std::vector<CTxIn> &vecToExclude, int nProtocolVersion = -1);
+    CMasternode* FindRandomNotInSet(std::set<CTxIn> &vecToExclude, int nProtocolVersion = -1);
```
Comment (UdjinM6) [inline src/masternodeman.h]: vecToExclude -> setToExclude

**Comment by crowning-** [comment]
As mentioned in the discussion here https://github.com/dashpay/dash/pull/1055#issuecomment-250941884, when you change this from `GetRandInt` to the standard `rand()` this should be perfect.

**Comment by tgflynn** [comment]
No, I think @crowning- 's PR is faster.  But maybe it's worth combining the two.

Is it safe to use rand() instead of GetRandInt (which I'm guessing is a cryptographic suitably unpredictable PRNG) here ?  I don't know enough about how PS works to know how important unpredictability is here.

**Comment by crowning-** [comment]
When you seed it with the current time you should be fine because the iteration time for each call is way > 1 sec.
[http://www.azillionmonkeys.com/qed/random.html](url), chapter "A More Reasonable Answer" might be needed, though,

**Comment by tgflynn** [comment]
Yeah but the time is pretty predictable.

I couldn't access that url for some reason.

**Comment by crowning-** [comment]
Does http://216.92.85.227/qed/random.html work?

**Comment by tgflynn** [comment]
Yes.

**Comment by tgflynn** [comment]
It looks like the current RNG: GetRandInt calls RAND_bytes which is an openssl function (see the man page).

This is probably using some system entropy source because there's another function RAND_pseudo_bytes.

Maybe using RAND_bytes on each call is overkill here.  Maybe we could just seed a PRNG from a single strong random number and then pass the PRNG to shuffle.

**Comment by crowning-** [comment]
> Maybe we could just seed a PRNG from a single strong random number and then pass the PRNG to shuffle.

Sounds perfect for our purpose :thumbsup:

**Comment by UdjinM6** [comment]
Ok, I did some benchmarks on mainnet.

Results:
- original code with modifications http://pastebin.com/CSCPVH99
- #1055 http://pastebin.com/xMK0WAub
- this PR http://pastebin.com/XSpcPnmX

`insecure_rand`/benchmark patches:
- #1055 http://pastebin.com/P8i0v2js
- this PR http://pastebin.com/cDxakk3C

Basically #1055 is no go, it's way too slow. And it's getting slower and slower when vector grows just as I thought.
This PR brings no significant improvement in performance comparing to original code when using fast `insecure_rand` instead of slow `GetRandInt`. Keeping in mind that here we use 2 mirrored data structures to keep the same data (which adds complexity) I think this PR is not worth it so I temporary revoke my ACK here.

Instead I propose to substitute `GetRandInt` with `insecure_rand` in original code and left the rest of the code as it is.

**Comment by tgflynn** [comment]
What size exclusion vector did you run these with ?

Your differences between the original and insecure_rand seem much smaller than crowning's factor of 10, though I think insecure_rand should be a very fast function.

**Comment by UdjinM6** [comment]
It's full mainnet vector - 3600+. Results are for unspecified (i.e. no rand function provided at all) and `insecure_rand`. Sorry, I wasn't clear about that. With `GetRandInt` for original code it is ~5-8ms, with few spikes up to 15ms. I haven't tested #1055 or this with `GetRandInt` but I guess it would be also ~3-5 times slower.

**Comment by evan82** [comment]
utACK

**Comment by UdjinM6** [comment]
@evan82 are you sure we need set/list instead of vector here? Check my benchmarks above, there is almost no difference in performance comparing to original code when both use `insecure_rand`.

**Comment by crowning-** [comment]
@evan82 @UdjinM6 I'll do some more tests today after work and report back:

- compare speed of the standard `rand()` and `insecure_rand` calls with the original (= using vectors) implementation. I'll upload a graph with the results
- investigate in the quality(!) of the created `insecure_rand` random numbers, especially how predictable they are. That's my main concern right now.

When that is done we can decide how to proceed.

**Comment by tgflynn** [comment]
I don't think the predictability of insecure_rand is a concern in this PR.  The PRNG is seeded in every call to FindRandomNotInSet with a secure, entropy based RNG.  Within the function it is only used to permute a vector so the values it returns are not externally observable.  The only observable effect of a call to the function is its choice of a single MN so there's nothing that can be predicted from that.

**Comment by tgflynn** [comment]
Thinking about my last comment I realized that we don't need to permute a vector at all.  New update coming.

**Comment by crowning-** [comment]
I guess the simplest solution is to just avoid shuffling _here_ at all and do it when a new Masternode is added to the main Masternode vector, because, besides the initial sync, this happens rarely compared to to mixing.

**Comment by tgflynn** [comment]
I don't think that works because what we need to be randomized is the set of non-excluded masternodes, not the full set.

One way is to just randomly pick one until we find one that's not excluded.  That would probably be faster than shuffling most of the time.  We could then fall back on shuffling if after a few tries we don't get a good MN.  I've implemented this but I want to do some timing tests before I submit it.

I think the best solution would be to rewrite CDarksendPool::DoAutomaticDenominating to not use this function at all but maintain it's own vector of non-excluded MN's from which it would would just pick one at random.  But that function is very complicated and this has to be done in a threadsafe way so it's definitely not something I would try to add to this PR.

**Comment by crowning-** [comment]
If our main Masternode vector is randomly filled, each subsequent search for a non-excluded Masternode has still a (globally) random result.

A simple
```
bool CMasternodeMan::Add(CMasternode &mn)
{
    LOCK(cs);

    if (!mn.IsEnabled() && !mn.IsPreEnabled())
        return false;

    CMasternode *pmn = Find(mn.vin);
    if (pmn == NULL)
    {
        LogPrint("masternode", "CMasternodeMan: Adding new Masternode %s - %i now\n", mn.addr.ToString(), size() + 1);

        // Insert new Masternodes into the vector at a random place
        unsigned long insertIndex = GetRandInt(vMasternodes.size());
        std::vector<CMasternode>::iterator mit = vMasternodes.begin() + insertIndex;
        printf("Insert %s at %lu of %lu\n", mn.addr.ToString().c_str(), insertIndex, vMasternodes.size());
        vMasternodes.insert(mit, mn);
        
        // vMasternodes.push_back(mn);
        return true;
    }

    return false;
}
```
(minus the printf) should do the job.

**Comment by crowning-** [comment]
And if there is never ever a new Masternode (could happen in theory), we could shuffle the vector when the exclusion vector is full, like `if(vecMasternodesUsed.size() > nThreshold) std::random_shuffle(vMasternodes.begin(), vMasternodes.end(), GetRandInt);

This doesn't happen very often, so it shouldn't hurt performance-wise.

**Comment by tgflynn** [comment]
I see two problems with this:

1) You still have to search the randomized vector for a non-excluded MN, so I'm not sure you really gain that much time.

2) You're always using the same ordering so I think it would make it much easier for an external observer to predict the next chosen MN.

**Comment by crowning-** [comment]
1) but we already agreed that the shuffling (more precise, the random number generator there) is the bottleneck. We don't need this any more
2) I'll re-order whenever the exclude-vector is full (see my last posting).

I'm running a PoC test right now, if it works I'll test the performance.

**Comment by UdjinM6** [comment]
I did some benchmarks removing shuffling and it's actually slower. I think that's because you always pick nodes from the start so they are most likely to be "used" without shuffling. So again, I'm proposing that we leave most of the code as it is simply replacing GetRandInt with insecure_rand.

**Comment by crowning-** [comment]
I agree, I did some measurements with a pre-shuffled vector of Masternodes and the times are around 0.01 ms per search, which is in the same ballpark as the existing code with `insecure_rand`:
findrandomnotinvec-getrandint-preorderedt

Note that I changed the deletion logic when the list of excluded Masternodes is full (that's the red line in the graph).
I'll put that independently of this PR into another PR tomorrow or so.

Who's gonna do the change to `insecure_rand` ?

**Comment by crowning-** [comment]
Nobody answered, so: https://github.com/dashpay/dash/pull/1063

**Comment by UdjinM6** [comment]
Thanks everyone! 👍 
Closing in fav of #1063

**Pull request closed without merging**